[Image] Add onLoadX support on Android

brentvatne opened the pull request:
~~This is a WIP, just finished the first bit and wanted to get some feedback to see if this approach seems appropriate, as I haven't done a lot of Android development.~~

Looks ready for review now.

mkonicek commented:
Cool! I'll pass this to Tyrone who works on Fresco to take a look.

mkonicek commented:
Should we make it consistent with iOS which has `onLoadStart`, `onLoadEnd` etc.?
https://facebook.github.io/react-native/docs/image.html

kmagiera commented on ReactAndroid/src/main/java/com/facebook/react/views/image/events/ImageLoadedEvent.java:
```diff
@@ -0,0 +1,40 @@
+
+  @Override
+  public void dispatch(RCTEventEmitter rctEventEmitter) {
+    rctEventEmitter.receiveEvent(getViewTag(), getEventName(), Arguments.createMap());
```
do you plan to actually pass some data here? Otherwise it would be more efficient to send null instead of creating empty map

brentvatne replied:
it looks like on iOS we just pass null, so unless someone can think of a good reason to do otherwise I'll change it to null here as well

kmagiera commented on ReactAndroid/src/main/java/com/facebook/react/views/image/ReactImageManager.java:
```diff
@@ -93,6 +103,33 @@ public void setTintColor(ReactImageView view, @Nullable Integer tintColor) {
+    final EventDispatcher mEventDispatcher = reactContext.getNativeModule(UIManagerModule.class)
+        .getEventDispatcher();
+
+    view.setLoadingControllerListener(new BaseControllerListener<ImageInfo>() {
```
I'd prefer to do sth similar to how we handle `onLayout` and only add that listener when `onLayout` is set in JS

brentvatne replied:
Makes sense

tyronen commented on ReactAndroid/src/main/java/com/facebook/react/views/image/ReactImageView.java:
```diff
@@ -212,6 +217,7 @@ public void maybeUpdateView() {
         .setOldController(getController())
         .setImageRequest(imageRequest)
         .setControllerListener(mControllerListener)
+        .setControllerListener(mLoadingControllerListener)
```
I'm afraid this will not work. Fresco only supports a single controller listener.

What you can do is define a single 'overall' controller listener that actually gets passed to Fresco. This internally would have a list of listeners. ReactImageView could then offer an addControllerListener method, and maybe also clearControllerListeners.

brentvatne replied:
Thanks! I'll do this

tyronen commented on ReactAndroid/src/main/java/com/facebook/react/views/image/ReactImageManager.java:
```diff
@@ -93,6 +103,33 @@ public void setTintColor(ReactImageView view, @Nullable Integer tintColor) {
+
+    view.setLoadingControllerListener(new BaseControllerListener<ImageInfo>() {
+      @Override
+      public void onFinalImageSet(
```
You can also override the `onSubmit` method to send an onLoadStart event, and `onFailure` should also send out onLoadEnd.

tyronen commented on ReactAndroid/src/main/java/com/facebook/react/views/image/events/ImageLoadedEvent.java:
```diff
@@ -0,0 +1,40 @@
+
+public class ImageLoadedEvent extends Event<ImageLoadedEvent> {
+
+  public static final String EVENT_NAME = "topImageLoad";
```
So as @mkonicek pointed out, iOS actually has five different image events. I'd looked into this earlier. Fresco can easily do onLoad, onLoadEnd, and onLoadStart. It can do onError some of the time. It can't yet do onProgress.

You should be able to pile all those types here. Don't use a Java enum, they're not efficient. Instead do this:

```java
​  @IntDef({ON_ERROR, ON_LOAD, ON_LOAD_END, ON_LOAD_START, ON_PROGRESS})
  @Retention(RetentionPolicy.SOURCE)
  @interface ImageEventType {}

  static final int ON_ERROR = 1;
  static final int ON_LOAD = 2;
  static final int ON_LOAD_END = 3;
  static final int ON_LOAD_START = 4;
  static final int ON_PROGRESS = 5;

  private final int mEventType;
```

tyronen commented on ReactAndroid/src/main/java/com/facebook/react/views/image/events/ImageLoadedEvent.java:
```diff
@@ -0,0 +1,40 @@
+
+  public static final String EVENT_NAME = "topImageLoad";
+
+  public ImageLoadedEvent(int viewId, long timestampMs) {
```
Add an int for the type to the constructor.

tyronen commented on ReactAndroid/src/main/java/com/facebook/react/views/image/events/ImageLoadedEvent.java:
```diff
@@ -0,0 +1,40 @@
+
+  @Override
+  public String getEventName() {
+    return EVENT_NAME;
```
Have a switch statement here based on the event type. You can return "topImageLoad", "topImageStart" etc.

brentvatne commented:
@mkonicek @tyronen @kmagiera - thanks for the great tips! I've factored them in and updated this PR. I also added an example to UIExplorer.

At the moment it supports: `onLoadStart`, `onLoadEnd` and `onLoad`. I added a comment about the absence of `onError` and `onProgress` being missing -- we can circle back and add those when support exists in fresco if that sounds good to you guys.

satya164 commented:
+1

kmagiera commented on Libraries/Image/Image.ios.js:
```diff
@@ -128,12 +127,10 @@ var Image = React.createClass({
     onError: PropTypes.func,
     /**
      * Invoked when load completes successfully
-     * @platform ios
```
you mentioned `onError` isn't working on android, no?

brentvatne replied:
correct -- this line is actually removing the platform annotation for `onLoad`

kmagiera commented on ReactAndroid/src/main/java/com/facebook/react/views/image/ReactImageManager.java:
```diff
@@ -92,6 +96,23 @@ public void setTintColor(ReactImageView view, @Nullable Integer tintColor) {
     }
   }
 
+  @ReactProp(name = "loadHandlersRegistered")
```
I'd prefer for this property to be called "shouldNotifyLoadEvents" or similar instead "loadHandlersRegistered"

kmagiera commented on ReactAndroid/src/main/java/com/facebook/react/views/image/ReactImageManager.java:
```diff
@@ -92,6 +96,23 @@ public void setTintColor(ReactImageView view, @Nullable Integer tintColor) {
   }
 
+  @ReactProp(name = "loadHandlersRegistered")
+  public void setloadHandlersRegistered(ReactImageView view, boolean shouldNotifyOnLoadEvents) {
```
`setload...` -> `setLoad`

kmagiera commented on ReactAndroid/src/main/java/com/facebook/react/views/image/ReactImageView.java:
```diff
@@ -123,10 +133,59 @@ public ReactImageView(
   }
 
+  public void setShouldNotifyOnLoadEvents(boolean shouldNotify) {
+    if (!shouldNotify || mHasRegisteredForOnLoadEvents) {
```
when `shouldNotify=false` you should remove the listener.

kmagiera commented on ReactAndroid/src/main/java/com/facebook/react/views/image/ReactImageView.java:
```diff
@@ -219,9 +278,7 @@ public void maybeUpdateView() {
 
   // VisibleForTesting
   public void setControllerListener(ControllerListener controllerListener) {
-    mControllerListener = controllerListener;
```
This is used in tests only. Just to handle this case you allocate extra `forwardingcontroller` instance in the constructor. Instead we should have two member refs: `mControllerForTesting` and `mControllerListener` then set it in this method and in `setShouldNotifyOnLoadEvents`  respectively and only in `maybeUpdateView` we can allocate `forwardingController` in case when both are non-null (which will only happen in our instrumentation tests)

brentvatne replied:
Fixed!

kmagiera commented on ReactAndroid/src/main/java/com/facebook/react/views/image/events/ImageLoadEvent.java:
```diff
@@ -0,0 +1,76 @@
+  @Override
+  public short getCoalescingKey() {
+    // All events for a given view can be coalesced.
+    return 0;
```
this may make `onLoadStart` and `onLoadEnd` be coalesced, as result JS will only receive `onLoadEnd`. This is definitely not sth we want as in JS someone may want to add logic that is triggered by "start" or rely on getting "start" before "end" events. You should use `eventType` for the coalescing key or alternatively have separate event class per event type (I prefer the former).

kmagiera commented:
Getting there - few more comments

mkonicek commented:
Thanks for working on this!

brentvatne commented:
@kmagiera - just on vacation atm so not as much time but I will try to get to this over the next few days :) thanks for the review!

mkonicek commented:
You're traveling across Europe and I saw you updated the ViewPager PR :) Enjoy vacation Brent :)

brentvatne commented:
@kmagiera - updated, squashed, rebased :rocket:

kmagiera commented on ReactAndroid/src/main/java/com/facebook/react/views/image/ReactImageView.java:
```diff
@@ -127,6 +137,56 @@ public ReactImageView(
+  public void setShouldNotifyLoadEvents(boolean shouldNotify) {
+    if (!shouldNotify) {
+      mControllerListener = null;
+      return;
```
I think you actually want to set `mIsDirty = true` when listener is unset. Otherwise calling this method with `shouldNotify = false` may have no effect

kmagiera commented on ReactAndroid/src/main/java/com/facebook/react/views/image/ReactImageView.java:
```diff
@@ -127,6 +137,56 @@ public ReactImageView(
+    }
+
+    mIsDirty = true;
+    maybeUpdateView();
```
This is unnecessary and may trigger additional work in case when multiple properties are updated in the same batch. This method should be only called from `onAfterUpdateTransation` which covers all the cases when properties has been modified https://github.com/facebook/react-native/blob/e4f0971a55d386db93ff5c78d383ffae1132dc33/ReactAndroid/src/main/java/com/facebook/react/views/image/ReactImageManager.java#L108

(we actually also call it from `onSizeChanged` but this may happen independently, therefore it's fine)

brentvatne replied:
Ah thanks for pointing this out, I think what confused me here was this: https://github.com/facebook/react-native/blob/e4f0971a55d386db93ff5c78d383ffae1132dc33/ReactAndroid/src/main/java/com/facebook/react/views/image/ReactImageView.java#L233-L237

kmagiera commented on ReactAndroid/src/main/java/com/facebook/react/views/image/ReactImageView.java:
```diff
@@ -217,21 +277,31 @@ public void maybeUpdateView() {
         .build();
 
-    DraweeController draweeController = mDraweeControllerBuilder
+    mDraweeControllerBuilder
```
nit: separate call to `.reset()` with a comment that the same builder is going to be reused, then continue with all the remaining calls

kmagiera commented on ReactAndroid/src/main/java/com/facebook/react/views/image/events/ImageLoadEvent.java:
```diff
@@ -0,0 +1,77 @@
+  @Retention(RetentionPolicy.SOURCE)
+  @interface ImageEventType {}
+
+  /**
```
`/**` is for javadoc block. This is not javadoc, we normally use single line comment for such a thing. E.g:

    // Currently ON_ERROR and ON_PROGRESS are not implemented, these can be added
    // easily once support exists in fresco.

kmagiera commented on ReactAndroid/src/main/java/com/facebook/react/views/image/events/ImageLoadEvent.java:
```diff
@@ -0,0 +1,77 @@
+  public void dispatch(RCTEventEmitter rctEventEmitter) {
+    rctEventEmitter.receiveEvent(getViewTag(), getEventName(), null);
+  }
+
```
kill this line

kmagiera commented:
Looks great. Just a few more things outlined above. Thanks for your patience

brentvatne commented on ReactAndroid/src/main/java/com/facebook/react/views/image/ReactImageView.java:
```diff
@@ -127,6 +137,56 @@ public ReactImageView(
+        public void onFinalImageSet(
+            String id,
+            @Nullable final ImageInfo imageInfo,
+            @Nullable Animatable animatable) {
```
as discussed on messenger we'll leave this one as is :)

brentvatne commented:
@kmagiera @mkonicek - feedback addressed, thanks guys!

kmagiera commented:
Great job, this PR deserves some cool gif
![](http://i.imgur.com/7qV8Otf.gif)

@facebook-github-bot shipit

mkonicek commented:
Failed to land due to a Flow error. You can also see a repro for that on Travis. I asked the Flow guys about the correct way to fix this.

brentvatne commented:
@mkonicek - fixed that

mkonicek commented:
@facebook-github-bot shipit

mkonicek commented:
Finally landed :tada: 

The Android PRs often require changes to Buck files in order for them to build internally. This makes merging them quite a bit of work :( Some ideas how to fix this:
- Have one Buck file for all of OSS RN instead of many small Buck files
- Open source all Buck files and report failures back to github (PITA)

vjeux commented on ReactAndroid/src/main/java/com/facebook/react/views/image/ReactImageView.java:
```diff
@@ -127,6 +137,48 @@ public ReactImageView(
+        }
+
+        @Override
+        public void onFailure(String id, Throwable throwable) {
```
You are saying that fresco doesn't support ON_ERROR but you have onFailure here. Is the comment not correct? Any reason why you don't throw ON_ERROR here?

vjeux commented on ReactAndroid/src/main/java/com/facebook/react/views/image/events/ImageLoadEvent.java:
```diff
@@ -0,0 +1,73 @@
+  @Retention(RetentionPolicy.SOURCE)
+  @interface ImageEventType {}
+
+  // Currently ON_ERROR and ON_PROGRESS are not implemented, these can be added
```
comment here

foghina commented on ReactAndroid/src/main/java/com/facebook/react/views/image/ReactImageView.java:
```diff
@@ -127,6 +137,48 @@ public ReactImageView(
+        @Override
+        public void onFailure(String id, Throwable throwable) {
+          mEventDispatcher.dispatchEvent(
+              new ImageLoadEvent(getId(), SystemClock.uptimeMillis(), ImageLoadEvent.ON_LOAD_END)
```
@brentvatne / @kmagiera sorry to be a necromancer :) but is there any particular reason why we're not using `ImageLoadEvent.ON_ERROR` here? And what is the note about fresco not supporting `ON_ERROR` below about, isn't `onFailure` pretty much the same thing?

brentvatne replied:
Reason was this: https://github.com/facebook/react-native/pull/3791#discussion_r43493004

vjeux commented on ReactAndroid/src/main/java/com/facebook/react/views/image/ReactImageView.java:
```diff
@@ -127,6 +137,48 @@ public ReactImageView(
+            mEventDispatcher.dispatchEvent(
+                new ImageLoadEvent(getId(), SystemClock.uptimeMillis(), ImageLoadEvent.ON_LOAD_END)
+            );
+            mEventDispatcher.dispatchEvent(
```
fyi, those two events are called in the reverse order on iOS
